driver-async/src/test/functional/com/mongodb/async/client/CrudTest.java

```diff
@@ -120,6 +130,34 @@ public void shouldPassAllOutcomes() {
         }
     }
 
+    private void runOperationV2(final BsonDocument expectedOutcome, final BsonDocument outcome, final BsonDocument expectedResult) {
+        if (expectedOutcome.containsKey("error")) {
+            assertEquals("Expected error", expectedOutcome.getBoolean("error"), outcome.get("error"));
+        }
+
+        if (expectedResult != null) {
+            // Hack to workaround the lack of upsertedCount
+            BsonValue actualResult = outcome.get("result");
+            if (actualResult.isDocument()
+                    && actualResult.asDocument().containsKey("upsertedCount")
+                    && actualResult.asDocument().getNumber("upsertedCount").intValue() == 0
+                    && !expectedResult.asDocument().containsKey("upsertedCount")) {
+                expectedResult.asDocument().append("upsertedCount", actualResult.asDocument().get("upsertedCount"));
+            }
+            // Hack to workaround the lack of insertedIds
+            if (expectedResult.isDocument()
+                    && !expectedResult.asDocument().containsKey("insertedIds")) {
+                actualResult.asDocument().remove("insertedIds");
+            }
+
+            assertEquals(description, expectedResult, actualResult);
+        }
+
+        if (expectedOutcome.containsKey("collection")) {
+            assertCollectionEquals(expectedOutcome.getDocument("collection"));
+        }
```

Comment:
Looks like there are also command events in the json: `command_started_event` these also need to be checked.